proposals/2024-04-03_multiple-exemplars-per-sample.md

```diff
@@ -0,0 +1,85 @@
+## Multiple exemplars per sample
+
+* **Owners:**
+  * György (krajo) Krajcsovits [@krajorama](https://github.com/krajorama/) [gyorgy.krajcsovits@grafana.com](mailto:gyorgy.krajcsovits@grafana.com)
+
+* **Implementation Status:** Not implemented
+
+* **Related Issues and PRs:**
+  * Prometheus [Support for ingesting out of order exemplars](https://github.com/prometheus/prometheus/issues/13577)
+  * Prometheus [WIP: tsdb: Add support for out-of-order exemplars](https://github.com/prometheus/prometheus/pull/13580)
+  * Prometheus client model [How to add exemplars to pure Native Histograms](https://github.com/prometheus/client_model/issues/61)
+  * Prometheus client golang [histograms: Support exemplars in pure Native Histograms](https://github.com/prometheus/client_golang/issues/1126)
+  * Prometheus client golang [add native histogram exemplar support](https://github.com/prometheus/client_golang/pull/1471)
+  * Prometheus [scrape: Enable ingestion of multiple exemplars per sample](https://github.com/prometheus/prometheus/pull/12557)
+  * Prometheus [Native histograms: error on ingesting out-of-order exemplars](https://github.com/prometheus/prometheus/issues/12971)
+  * Prometheus [Fix error on ingesting out-of-order exemplars](https://github.com/prometheus/prometheus/pull/13021)
+  * Prometheus [Prometheus support parse exemplars from native histogram](https://github.com/prometheus/prometheus/pull/13488)
+  * OTEL proto [Base-2 exponential histogram protocol support](https://github.com/open-telemetry/opentelemetry-proto/pull/322)
+
+* **Other docs or links:**
+  * [OTEL histogram](https://opentelemetry.io/docs/specs/otel/metrics/data-model/#histogram)
+  * [OTEL exemplars](https://opentelemetry.io/docs/specs/otel/metrics/data-model/#exemplars)
+
+> This proposal collects the requirements and implementation proposals for supporting multiple exemplars per sample, in particular for native histograms.
+
+## Why
+
+Currently the remote write protocol (1.0 and 2.0) processes exemplars individually, which means that if a native histogram has multiple exemplars, they may arrive in separate remote write requests, making it hard to decide if the exemplar is out of order it belonged to a previous set of exemplars and should be considered a duplicate.
+
+Currently there is no mapping of the "set of exemplars" per sample concept that is in Open Telemetry specification.
+
+### Pitfalls of the current solution
+
+In the current solution we have a circular buffer for all exemplars. The exemplars are organized into two lists. One list links all exemplars from oldest received to newest so that we can make space when maximum number of exemplars is hit on insert. The other link
+is per time series, linking exemplars from oldest to newest. Also there is a pointer to the latest exemplar per time series to be able to check for duplicates, out of order.
```

Comment:
Do you mean "the other list" rather than "the other link"?
```suggestion
In the current solution we have a circular buffer for all exemplars. The exemplars are organized into two lists. One list links all exemplars from oldest received to newest so that we can make space when maximum number of exemplars is hit on insert. The other list
is per time series, listing exemplars from oldest to newest. Also there is a pointer to the latest exemplar per time series to be able to check for duplicates, out of order.
```